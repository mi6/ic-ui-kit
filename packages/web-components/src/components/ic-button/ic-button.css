@import "../../global/normalize.css";

/**
* @prop --min-width: Minimum width of the button.
*/

:host {
  display: inline-block;

  --button-default: var(--ic-action-default);
  --button-default-hover: var(--ic-action-default-hover);
  --button-default-active: var(--ic-action-default-active);
  --button-default-background-hover: var(--ic-action-default-bg-hover);
  --button-default-background-active: var(--ic-action-default-bg-active);
  --icon-width: 100%;
  --icon-height: 100%;
}

.button {
  font-family: var(--ic-font-body-family);
  text-decoration: none;
  font-weight: 600;
  font-size: 0.875rem;
  transition: var(--ic-easing-transition-fast);
  border-radius: var(--ic-border-radius);
  min-width: var(--min-width, 6.25rem);
  display: inline-flex;
  gap: var(--ic-space-xs);
  flex-direction: row;
  justify-content: center;
  align-items: center;
  background: none;
  border: none;
  box-sizing: border-box;
  white-space: nowrap;
  vertical-align: middle;
}

.button:hover {
  cursor: pointer;
}

.button:focus,
:host .button:focus {
  box-shadow: var(--ic-border-focus);
}

:host(.light) .button:focus {
  box-shadow: var(--ic-border-focus);
}

.button:focus-visible {
  outline: var(--ic-hc-focus-outline);
}

:host(.dark) .button {
  --button-default: var(--ic-action-dark);
  --button-default-hover: var(--ic-action-dark);
  --button-default-active: var(--ic-action-dark);
  --button-default-background-hover: var(--ic-action-dark-bg-hover);
  --button-default-background-active: var(--ic-action-dark-bg-active);
}

:host(.light) .button {
  --button-default: var(--ic-action-light);
  --button-default-hover: var(--ic-action-light);
  --button-default-active: var(--ic-action-light);
  --button-default-background-hover: var(--ic-action-light-bg-hover);
  --button-default-background-active: var(--ic-action-light-bg-active);
}

:host(.button-variant-primary.light) .button {
  color: var(--ic-color-primary-text);

  --button-default-hover: var(--ic-action-light-hover);
  --button-default-active: var(--ic-action-light-active);
}

:host(.button-variant-primary.dark) .button {
  --button-default-hover: var(--ic-action-dark-hover);
  --button-default-active: var(--ic-action-dark-active);
}

:host(.disabled),
:host(.disabled) .button,
:host(.loading),
:host(.loading) .button {
  pointer-events: none;
}

:host(.loading-with-icon) .button {
  min-width: var(--min-width, 8.25rem);
}

/* Variants */

/* Primary */

:host(.button-variant-primary) .button {
  color: var(--ic-architectural-white);
  background-color: var(--button-default);
}

:host(.button-variant-primary) .button:hover:not(:focus) {
  background-color: var(--button-default-hover);
}

:host(.button-variant-primary) .button:active:not(:focus),
:host(.button-variant-primary.loading) .button {
  background-color: var(--button-default-active);
}

:host(.button-variant-primary.disabled) .button {
  background: var(--ic-architectural-200);
  color: var(--ic-architectural-300);
}

:host(.button-variant-primary.light.disabled) .button {
  background: var(--ic-architectural-600);
  color: var(--ic-architectural-500);
}

/* Secondary */

:host(.button-variant-secondary) .button {
  border: var(--ic-space-1px) solid var(--button-default);
  color: var(--button-default);
}

:host(.button-variant-secondary) .button:hover:not(:focus) {
  background-color: var(--button-default-background-hover);
  border-color: var(--button-default-hover);
  color: var(--button-default-hover);
}

:host(.button-variant-secondary) .button:active:not(:focus) {
  border-color: var(--button-default-active);
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-secondary.loading) .button {
  border-color: var(--button-default-active);
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-secondary.disabled) .button,
:host(.button-variant-secondary.disabled) .button:hover,
:host(.button-variant-secondary.disabled) .button:active {
  border-color: var(--ic-architectural-300);
  color: var(--ic-architectural-300);
  background: none;
}

:host(.button-variant-secondary.light.disabled) .button,
:host(.button-variant-secondary.light.disabled) .button:hover,
:host(.button-variant-secondary.light.disabled) .button:active {
  border-color: var(--ic-architectural-500);
  color: var(--ic-architectural-500);
  background: none;
}

/* Tertiary */

:host(.button-variant-tertiary) .button {
  color: var(--button-default);
}

:host(.button-variant-tertiary) .button:hover:not(:focus) {
  background-color: var(--button-default-background-hover);
  color: var(--button-default-hover);
}

:host(.button-variant-tertiary) .button:active:not(:focus),
:host(.button-variant-tertiary.loading) .button {
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-tertiary.disabled) .button,
:host(.button-variant-tertiary.disabled) .button:hover,
:host(.button-variant-tertiary.disabled) .button:active {
  border-color: var(--ic-architectural-300);
  color: var(--ic-architectural-300);
  background: none;
}

:host(.button-variant-tertiary.light.disabled) .button,
:host(.button-variant-tertiary.light.disabled) .button:hover,
:host(.button-variant-tertiary.light.disabled) .button:active {
  border-color: var(--ic-architectural-500);
  color: var(--ic-architectural-500);
  background: none;
}

/* Icon */

:host(.button-variant-icon) .button {
  color: var(--button-default);
  min-width: 0;
}

:host(.button-variant-icon) .button:hover:not(:focus) {
  background-color: var(--button-default-background-hover);
  color: var(--button-default-hover);
}

:host(.button-variant-icon) .button:active:not(:focus),
:host(.button-variant-icon.loading) .button {
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-icon.disabled) .button,
:host(.button-variant-icon.disabled) .button:hover,
:host(.button-variant-icon.disabled) .button:active {
  color: var(--ic-architectural-300);
  background: none;
}

/* Destructive */

:host(.button-variant-destructive) .button {
  color: var(--ic-architectural-white);
  background-color: var(--ic-action-destructive);
  text-transform: uppercase;
}

:host(.button-variant-destructive) .button:hover:not(:focus) {
  background-color: var(--ic-action-destructive-hover);
}

:host(.button-variant-destructive) .button:active:not(:focus),
:host(.button-variant-destructive.loading) .button {
  background-color: var(--ic-action-destructive-active);
}

:host(.button-variant-destructive.disabled) .button {
  background: var(--ic-architectural-200);
  color: var(--ic-architectural-300);
}

/* Sizing */

:host(.button-size-default) .button {
  height: 2.5rem;
  padding: var(--ic-space-xs) var(--ic-space-md);
}

:host(.button-size-small) .button {
  height: var(--ic-space-xl);
  padding: var(--ic-space-xxs) var(--ic-space-md);
}

:host(.button-size-large) .button {
  height: var(--ic-space-xxl);
  padding: var(--ic-space-sm) var(--ic-space-md);
}

:host(.button-size-default.button-variant-icon) .button {
  height: var(--ic-space-xl);
  width: var(--ic-space-xl);
  padding: 0.375rem;
}

:host(.button-size-small.button-variant-icon) .button {
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
  padding: var(--ic-space-xxs);
}

:host(.button-size-large.button-variant-icon) .button {
  height: 2.5rem;
  width: 2.5rem;
  padding: var(--ic-space-xs);
}

/* Width */
:host(.full-width),
:host(.full-width) .button {
  width: 100%;
}

/* Loading */

div.loading-container {
  position: relative;
  align-items: center;
  width: 100%;
}

ic-loading-indicator {
  --outer-color: transparent;
}

@keyframes loading-animation {
  0% {
    width: 0%;
    left: 0%;
  }

  25% {
    width: 0%;
    left: 0%;
    opacity: 0;
  }

  50% {
    width: 100%;
    left: 0%;
    opacity: 1;
  }

  75% {
    width: 0%;
    left: 100%;
  }

  100% {
    width: 0%;
    left: 100%;
    opacity: 0;
  }
}

/* Icons */

div.icon-container {
  box-sizing: border-box;
  width: var(--ic-space-lg);
  height: var(--ic-space-lg);
  display: flex;
  justify-content: center;
  align-items: center;
}

::slotted(*) {
  width: var(--icon-width) !important;
  height: var(--icon-height) !important;
  fill: currentcolor !important;
  pointer-events: none;
}

:host(.button-variant-icon) .button .icon-container {
  margin: 0;
  pointer-events: none;
}

/** SEARCH **/

:host(.search-submit-button) ::slotted(svg) {
  --icon-height: 1.25rem;
  --icon-width: 1.25rem;
}

:host(.search-submit-button-small) ::slotted(svg) {
  --icon-height: 1rem;
  --icon-width: 1rem;
}

:host(.clear-button) {
  margin: 0 var(--ic-space-xxs);
}

:host(.clear-button) .button:focus {
  box-shadow: none;
}

:host(.search-submit-button) .button:focus {
  box-shadow: none;
}

:host(.search-submit-button) .button:not(:active):focus {
  box-shadow: none;
  background-color: var(--ic-action-default-bg-hover);
}

:host(.search-submit-button) {
  display: flex;
  align-items: center;
  margin: 0 var(--ic-space-xxs);
}

:host(.menu-close-button) ::slotted(svg) {
  --icon-height: 0.875rem;
  --icon-width: 0.875rem;
}

:host(.popout-menu-button) .button {
  justify-content: left;
  border-radius: 0;
  white-space: pre-line;
  text-align: start;
}

:host(.popout-menu-button) div.icon-container {
  flex: none;
}

:host(.popout-menu-button) .button:focus {
  box-shadow: var(--ic-border-focus-inset);
  border-radius: var(--ic-border-radius-inset);
}

.ic-button-describedby {
  display: none;
}

:host .ic-tooltip {
  display: block;
}

/** High Contrast **/
@media (forced-colors: active) {
  .button {
    border: 0.125rem solid transparent;
  }
}

/** Pagination **/

:host(.flip) ::slotted(svg) {
  transform: scaleX(-1);
}
