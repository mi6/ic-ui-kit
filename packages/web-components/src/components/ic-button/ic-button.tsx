import {
  Component,
  Element,
  Event,
  EventEmitter,
  Host,
  Listen,
  Method,
  Prop,
  State,
  h,
} from "@stencil/core";

import {
  getThemeFromContext,
  inheritAttributes,
  isSlotUsed,
  removeDisabledFalse,
} from "../../utils/helpers";
import { IC_INHERITED_ARIA } from "../../utils/constants";
import {
  IcButtonTypes,
  IcButtonVariants,
  IcButtonTooltipPlacement,
} from "./ic-button.types";
import {
  IcTheme,
  IcThemeForeground,
  IcThemeForegroundEnum,
  IcSizes,
} from "../../utils/types";

let buttonIds = 0;

/**
 * @slot icon - Content will be placed to the left of the button label.
 */
@Component({
  tag: "ic-button",
  styleUrl: "ic-button.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class Button {
  private buttonEl: HTMLElement;
  private buttonIdNum = buttonIds++;
  private hasTooltip: boolean = false;
  private id: string;
  private inheritedAttributes: { [k: string]: unknown } = {};
  private tooltipEl: HTMLIcTooltipElement;
  private describedByEl: HTMLElement = null;
  private describedById: string = null;
  private mutationObserver: MutationObserver = null;

  @Element() el: HTMLIcButtonElement;

  @State() describedByContent: string = null;

  /**
   * The appearance of the button, e.g. dark, light, or the default.
   */
  @Prop({ mutable: true }) appearance?: IcThemeForeground = "default";

  /**
   * If `true`, the button will be in disabled state.
   */
  @Prop() disabled?: boolean = false;

  /**
   * If `true`, the ic-tooltip which is shown for icon variant will be disabled. Title or aria-label must be set if this prop is not applied.
   */
  @Prop() disableTooltip?: boolean = false;

  /**
   * If `true`, the user can save the linked URL instead of navigating to it.
   */
  @Prop() download?: string | boolean = false;

  /**
   * The <form> element to associate the button with.
   */
  @Prop() form?: string;

  /**
   * The URL that processes the information submitted by the button. It overrides the action attribute of the button's form owner. Does nothing if there is no form owner.
   */
  @Prop() formaction?: string;

  /**
   * The way the submitted form data is encoded.
   */
  @Prop() formenctype?: string;

  /**
   * The HTTP method used to submit the form.
   */
  @Prop() formmethod?: string;

  /**
   * If `true`, the form will not be validated when submitted.
   */
  @Prop() formnovalidate?: boolean;

  /**
   * The place to display the response from submitting the form. It overrides the target attribute of the button's form owner.
   */
  @Prop() formtarget?: string;

  /**
   * If `true`, the button will fill the width of the container.
   */
  @Prop() fullWidth?: boolean = false;

  /**
   * The URL that the link points to. This will render the button as an "a" tag.
   */
  @Prop() href?: string;

  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string;

  /**
   * If `true`, the button will be in loading state.
   */
  @Prop() loading?: boolean = false;

  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;

  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;

  /**
   * The size of the button to be displayed.
   */
  @Prop() size?: IcSizes = "default";

  /**
   * The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;

  /**
   * The position of the tooltip in relation to the button.
   */
  @Prop() tooltipPlacement?: IcButtonTooltipPlacement = "bottom";

  /**
   * The type of the button.
   */
  @Prop() type?: IcButtonTypes = "button";

  /**
   * The variant of the button to be displayed.
   */
  @Prop() variant?: IcButtonVariants = "primary";

  /**
   * Emitted when button has blur
   */
  @Event() icBlur!: EventEmitter<void>;

  /**
   * Emitted when button has focus
   */
  @Event() icFocus!: EventEmitter<void>;

  disconnectedCallback(): void {
    if (this.mutationObserver !== null && this.mutationObserver !== undefined) {
      this.mutationObserver.disconnect();
    }
  }

  componentWillLoad(): void {
    this.inheritedAttributes = inheritAttributes(this.el, [
      ...IC_INHERITED_ARIA,
      "aria-expanded",
      "title",
    ]);

    removeDisabledFalse(this.disabled, this.el);

    this.el.setAttribute("exportparts", "button");

    const id = this.el.id;
    this.id = id !== undefined ? id : null;
    this.hasTooltip = this.variant === "icon" && this.disableTooltip === false;

    if (!this.hasTooltip) {
      const describedById = this.inheritedAttributes[
        "aria-describedby"
      ] as string;
      if (describedById !== undefined) {
        this.describedById = describedById;
        const el = this.el.parentElement.querySelector(
          `#${describedById}`
        ) as HTMLElement;
        this.describedByContent = el.innerText;
        this.describedByEl = el;
      }
    }
  }

  componentDidLoad(): void {
    this.updateTheme();

    if (this.describedById) {
      this.mutationObserver = new MutationObserver(this.mutationCallback);
      this.mutationObserver.observe(this.describedByEl, {
        characterData: true,
        childList: true,
        subtree: true,
      });
    }
  }

  @Listen("click", { capture: true })
  handleHostClick(event: Event): void {
    if (this.disabled || this.loading) {
      event.stopImmediatePropagation();
    }
  }

  @Listen("themeChange", { target: "document" })
  themeChangeHandler(ev: CustomEvent): void {
    const theme: IcTheme = ev.detail;
    this.updateTheme(theme.mode);
  }

  /**
   * Sets focus on the native `button`.
   */
  @Method()
  async setFocus(): Promise<void> {
    if (this.buttonEl) {
      this.buttonEl.focus();
    }
  }

  /**
   * @internal Updates tooltip/aria-label text - needed as can't watch an ARIA attribute change.
   */
  @Method()
  async updateAriaLabel(newValue: string): Promise<void> {
    if (this.hasTooltip) {
      this.tooltipEl.label = newValue;
      this.buttonEl.removeAttribute("aria-label");
    } else {
      this.buttonEl.setAttribute("aria-label", newValue);
    }
  }

  private hasIconSlot(): boolean {
    const iconEl = this.el.querySelector(`[slot="icon"]`);
    return iconEl !== null;
  }

  private handleHiddenFormButtonClick(form: HTMLFormElement): void {
    const hiddenFormButton = document.createElement("button");

    hiddenFormButton.setAttribute("type", this.el.type);
    hiddenFormButton.style.display = "none";

    form.appendChild(hiddenFormButton);

    hiddenFormButton.click();
    hiddenFormButton.remove();
  }

  private handleClick = (): void => {
    if (
      (this.el.type === "submit" || this.el.type === "reset") &&
      !!this.el.closest("FORM")
    ) {
      this.handleHiddenFormButtonClick(this.el.closest("FORM"));
    }
  };

  private onFocus = () => {
    this.icFocus.emit();
  };

  private onBlur = () => {
    this.icBlur.emit();
  };

  private updateTheme(newTheme: IcThemeForeground = null): void {
    const foregroundColor = getThemeFromContext(this.el, newTheme || null);

    if (foregroundColor !== IcThemeForegroundEnum.Default) {
      this.appearance = foregroundColor;
    }
  }

  // triggered when text content of sibling element in light DOM changes
  private mutationCallback = (): void => {
    this.describedByContent = this.describedByEl.innerText;
  };

  render() {
    const TagType = (this.href && "a") || "button";
    const {
      title,
      "aria-label": ariaLabel,
      ...restInheritedAttributes
    } = this.inheritedAttributes;
    const buttonAttrs =
      TagType === "button"
        ? {
            type: this.type,
            disabled: this.disabled,
            form: this.form,
            formaction: this.formaction,
            formenctype: this.formenctype,
            formmethod: this.formmethod,
            formnovalidate: this.formnovalidate,
            formtarget: this.formtarget,
          }
        : {
            download: this.download !== false ? this.download : null,
            href: this.href,
            rel: this.rel,
            target: this.target,
            referrerpolicy: this.referrerpolicy,
            hreflang: this.hreflang,
          };
    const newTitle = title && (title as string);
    const titleAttr = this.hasTooltip ? {} : { title: newTitle };
    let tooltipText = "";
    if (this.hasTooltip) {
      if (newTitle !== undefined) {
        tooltipText = newTitle;
      } else if (ariaLabel !== null) {
        tooltipText = ariaLabel as string;
      }
    }

    let describedBy: string = null;
    let buttonId: string = null;
    if (this.hasTooltip) {
      buttonId =
        this.id !== null
          ? `ic-button-with-tooltip-${this.id}`
          : `ic-button-with-tooltip-${this.buttonIdNum}`;
      describedBy = `ic-tooltip-${buttonId}`;
    } else {
      describedBy = this.describedById;
    }

    const ButtonContent = () => {
      return (
        <TagType
          class="button"
          aria-disabled={this.loading || this.disabled ? "true" : null}
          aria-label={this.loading ? "Loading" : ariaLabel}
          {...buttonAttrs}
          {...restInheritedAttributes}
          {...titleAttr}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          ref={(el) => (this.buttonEl = el)}
          id={buttonId}
          aria-describedby={this.hasTooltip && ariaLabel ? null : describedBy}
          part="button"
        >
          {this.hasIconSlot() && !this.loading && (
            <div class="icon-container">
              <slot name="icon" />
            </div>
          )}
          {this.loading ? (
            <div class="loading-container">
              <ic-loading-indicator
                type="linear"
                appearance={
                  this.variant === "primary" ||
                  this.variant === "destructive" ||
                  this.appearance === IcThemeForegroundEnum.Dark ||
                  this.appearance === IcThemeForegroundEnum.Light
                    ? "light"
                    : "dark"
                }
              ></ic-loading-indicator>
            </div>
          ) : (
            <slot />
          )}
        </TagType>
      );
    };

    return (
      <Host
        class={{
          ["disabled"]: this.disabled && !this.loading,
          [`button-variant-${this.variant}`]: true,
          [`button-size-${this.size}`]: true,
          ["loading"]: this.loading,
          ["loading-with-icon"]: this.loading && this.hasIconSlot(),
          ["dark"]: this.appearance === IcThemeForegroundEnum.Dark,
          ["light"]: this.appearance === IcThemeForegroundEnum.Light,
          ["full-width"]: this.fullWidth,
          ["with-badge"]: isSlotUsed(this.el, "badge"),
        }}
        onClick={this.handleClick}
      >
        {this.hasTooltip && (
          <ic-tooltip
            class={{ ["tooltip-disabled"]: this.disableTooltip }}
            ref={(el) => (this.tooltipEl = el)}
            label={tooltipText}
            target={buttonId}
            placement={this.tooltipPlacement}
          >
            <ButtonContent />
          </ic-tooltip>
        )}

        {!this.hasTooltip && <ButtonContent />}
        {this.describedByContent && (
          <span id={describedBy} class="ic-button-describedby">
            {this.describedByContent}
          </span>
        )}
      </Host>
    );
  }
}
