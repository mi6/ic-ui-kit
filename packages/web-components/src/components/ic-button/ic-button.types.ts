export type IcButtonTypes = "reset" | "submit" | "button";

export type IcButtonVariants =
  | "primary"
  | "secondary"
  | "tertiary"
  | "icon"
  | "destructive";

export type IcButtonTooltipPlacement = "top" | "right" | "bottom" | "left";
