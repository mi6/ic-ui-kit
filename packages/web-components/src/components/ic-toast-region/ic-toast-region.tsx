import { Component, Element, h, Listen, Method } from "@stencil/core";
import { IcFocusableComponents } from "../../utils/types";

@Component({ tag: "ic-toast-region" })
export class ToastRegion {
  private pendingVisibility: HTMLIcToastElement[] = [];
  private previouslyFocused: HTMLElement;

  @Element() el: HTMLIcToastRegionElement;

  @Listen("icDismiss", { capture: true })
  handleDismissedToast(): void {
    if (this.pendingVisibility.length > 0) {
      this.pendingVisibility[0]
        .setVisible()
        .then((res) => (this.previouslyFocused = res));
      this.pendingVisibility.shift();
    } else {
      if (this.previouslyFocused && "setFocus" in this.previouslyFocused) {
        (this.previouslyFocused as IcFocusableComponents).setFocus();
      } else this.previouslyFocused?.focus();
    }
  }

  /**
   * Handles setting the visibility of various toasts based on what is already visible
   * @param toast The toast element being requested to display
   */
  @Method()
  async setVisible(toast: HTMLIcToastElement): Promise<void> {
    const visibleToasts = Array.from(
      document.querySelectorAll("ic-toast")
    ).filter((el) => window.getComputedStyle(el).display !== "none");
    if (visibleToasts.indexOf(toast) === -1 && visibleToasts.length <= 0) {
      toast.setVisible().then((res) => (this.previouslyFocused = res));
    }
    if (visibleToasts.length > 0) this.pendingVisibility.push(toast);
  }

  render() {
    return <slot></slot>;
  }
}
