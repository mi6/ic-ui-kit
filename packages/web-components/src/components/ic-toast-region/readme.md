# ic-toast-region



<!-- Auto Generated Below -->


## Methods

### `setVisible(toast: HTMLIcToastElement) => Promise<void>`

Handles setting the visibility of various toasts based on what is already visible

#### Returns

Type: `Promise<void>`




----------------------------------------------


