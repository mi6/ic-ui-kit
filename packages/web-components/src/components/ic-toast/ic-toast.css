@import "../../global/normalize.css";

/**
 * @prop --ic-z-index-toast: z-index of toast
 */

:host {
  --bottom-position: var(--ic-space-xl);

  position: fixed;
  bottom: var(--bottom-position);
  left: 50%;
  transform: translate(-50%);
  animation: fadein var(--ic-transition-duration-slow) ease-in-out;
  z-index: var(--ic-z-index-toast);
}

:host(.hidden) {
  display: none;
  animation: fadeout var(--ic-transition-duration-slow) ease-in-out;
}

.container {
  max-width: 32.5rem;
  min-width: 18rem;
  box-shadow: var(--ic-elevation-overlay);
  min-height: 3.5rem;
  background-color: var(--ic-architectural-800);
  color: var(--ic-color-white-text);
  display: flex;
  align-items: center;
  position: relative;
  border-radius: var(--ic-border-radius);
}

.toast-icon-container {
  height: 100%;
  display: flex;
  align-items: center;
}

.divider {
  height: 100%;
  width: var(--ic-space-xs);
  position: absolute;
  border-radius: var(--ic-space-xxxs) 0 0 var(--ic-space-xxxs);
}

.divider-neutral {
  background-color: var(--ic-status-unknown);
}

.divider-info {
  background-color: var(--ic-status-info-contrast);
}

.divider-warning {
  background-color: var(--ic-status-warning-contrast);
}

.divider-error {
  background-color: var(--ic-status-error-contrast);
}

.divider-success {
  background-color: var(--ic-status-success-contrast);
}

.toast-icon,
::slotted(svg) {
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
  margin-left: var(--ic-space-md);
}

:host([variant="neutral"]) .toast-icon svg,
::slotted(svg) {
  fill: var(--ic-status-unknown);
}

:host([variant="info"]) .toast-icon svg {
  fill: var(--ic-status-info-contrast);
}

:host([variant="warning"]) .toast-icon svg {
  fill: var(--ic-status-warning-contrast);
}

:host([variant="error"]) .toast-icon svg {
  fill: var(--ic-status-error-contrast);
}

:host([variant="success"]) .toast-icon svg {
  fill: var(--ic-status-success-contrast);
}

.toast-content {
  margin-left: var(--ic-space-xs);
  width: 100%;
}

.no-icon {
  margin-left: var(--ic-space-md);
}

.toast-message {
  padding: var(--ic-space-xs) var(--ic-space-xs) var(--ic-space-xs) 0;
}

.toast-action-container {
  padding-bottom: var(--ic-space-md);
}

ic-button,
.toast-dismiss-timer {
  padding: var(--ic-space-xs);
}

@media (max-width: 576px) {
  :host {
    --bottom-position: var(--ic-space-lg);

    width: calc(100% - 2 * var(--ic-space-md));
  }
}

@media (min-width: 993px) {
  :host {
    --bottom-position: calc(var(--ic-space-xl) + var(--ic-space-xs));
  }
}

@media (forced-colors: active) {
  .container {
    border: var(--ic-hc-border);
  }
}

@keyframes fadein {
  from {
    bottom: 0;
    opacity: 0;
  }

  to {
    bottom: var(--bottom-position);
    opacity: 1;
  }
}

@keyframes fadeout {
  from {
    bottom: var(--bottom-position);
    opacity: 1;
  }

  to {
    bottom: 0;
    opacity: 0;
  }
}
