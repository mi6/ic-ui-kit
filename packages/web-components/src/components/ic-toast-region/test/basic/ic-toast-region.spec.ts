import { newSpecPage } from "@stencil/core/testing";
import { Button } from "../../../ic-button/ic-button";
import { Toast } from "../../../ic-toast/ic-toast";
import { ToastRegion } from "../../ic-toast-region";

describe("ic-toast-region component", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [ToastRegion, Toast],
      html: `<ic-toast-region>
      <ic-toast heading="Heading"></ic-toast>
      </ic-toast-region>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should test showing and hiding toasts", async () => {
    const page = await newSpecPage({
      components: [ToastRegion, Toast],
      html: `<ic-toast-region>
      <ic-toast heading="Heading"></ic-toast>
      <ic-toast heading="Heading"></ic-toast>
      </ic-toast-region>`,
    });

    const toasts = page.root.querySelectorAll("ic-toast");

    await page.root.setVisible(toasts[0]);

    expect(page.rootInstance.pendingVisibility.length).toBe(1);

    await page.root.setVisible(toasts[1]);

    expect(page.rootInstance.pendingVisibility.length).toBe(2);

    await page.rootInstance.handleDismissedToast();
    expect(page.rootInstance.pendingVisibility.length).toBe(1);

    await page.rootInstance.handleDismissedToast();
    expect(page.rootInstance.pendingVisibility.length).toBe(0);

    await page.rootInstance.handleDismissedToast();
    expect(page.rootInstance.pendingVisibility.length).toBe(0);
  });

  it("should test previouslyFocused", async () => {
    const page = await newSpecPage({
      components: [ToastRegion, Toast, Button],
      html: `<ic-toast-region>
      <ic-button>Click Me</ic-button>
      <ic-toast heading="Heading"></ic-toast>
      </ic-toast-region>`,
    });

    const button = page.root.querySelector("ic-button");
    page.rootInstance.previouslyFocused = button;

    await page.rootInstance.handleDismissedToast();
    expect(page.rootInstance.pendingVisibility.length).toBe(0);
  });

  it("should test previously focused - standard element", async () => {
    const page = await newSpecPage({
      components: [ToastRegion, Toast],
      html: `<ic-toast-region>
      <button>Click Me</button>
      <ic-toast heading="Heading"></ic-toast>
      </ic-toast-region>`,
    });

    const button = page.root.querySelector("button");
    page.rootInstance.previouslyFocused = button;

    await page.rootInstance.handleDismissedToast();
    expect(page.rootInstance.pendingVisibility.length).toBe(0);
  });
});
