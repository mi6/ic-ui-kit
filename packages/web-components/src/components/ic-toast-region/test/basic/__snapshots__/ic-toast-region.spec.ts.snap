// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-toast-region component should render 1`] = `
<ic-toast-region>
  <!---->
  <ic-toast aria-label="Heading" class="ic-toast-hidden" heading="Heading" role="dialog" tabindex="0">
    <mock:shadow-root>
      <div class="container">
        <div class="toast-content">
          <div class="toast-text">
            <ic-typography class="toast-heading" variant="subtitle-large"></ic-typography>
          </div>
        </div>
        <ic-button aria-label="dismiss" id="dismiss-button" variant="icon">
          svg
        </ic-button>
      </div>
    </mock:shadow-root>
  </ic-toast>
</ic-toast-region>
`;
